File: python/aps/validate.py
"""JSON Schema validation for APS documents."""

import json
from pathlib import Path
from typing import Any

from jsonschema import Draft202012Validator, ValidationError  # noqa: F401

_SPEC_DIR = Path(__file__).resolve().parent.parent.parent / "spec"

def _load_schema(name: str) -> dict:
    with open(_SPEC_DIR / name) as f:
        return json.load(f)

def _validate(data: dict[str, Any], schema_name: str) -> None:
    schema = _load_schema(schema_name)
    validator = Draft202012Validator(schema)
    errors = sorted(validator.iter_errors(data), key=lambda e: list(e.path))
    if errors:
        raise ValidationError(errors[0].message, path=errors[0].path, schema_path=errors[0].schema_path)

def validate_passport(data: dict[str, Any]) -> None:
    """Validate a passport dict against agent-passport.schema.json. Raises ValidationError."""
    _validate(data, "agent-passport.schema.json")

def validate_receipt(data: dict[str, Any]) -> None:
    """Validate a receipt dict against work-receipt.schema.json. Raises ValidationError."""
    _validate(data, "work-receipt.schema.json")

def validate_envelope(data: dict[str, Any]) -> None:
    """Validate an envelope dict against security-envelope.schema.json. Raises ValidationError."""
    _validate(data, "security-envelope.schema.json")

def validate_dna(data: dict[str, Any]) -> None:
    """Validate a DNA dict against dna.schema.json. Raises ValidationError."""
    _validate(data, "dna.schema.json")

File: typescript/src/validate.ts
import Ajv2020_ from 'ajv/dist/2020.js';
const Ajv2020 = (Ajv2020_ as any).default ?? Ajv2020_;
import { readFileSync } from 'fs';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const specDir = resolve(__dirname, '../../spec');

function loadSchema(name: string): object {
  return JSON.parse(readFileSync(resolve(specDir, name), 'utf-8'));
}

export class ValidationError extends Error {
  constructor(message: string, public errors?: unknown[]) {
    super(message);
    this.name = 'ValidationError';
  }
}

function makeValidator(schemaName: string) {
  const schema = loadSchema(schemaName);
  const ajv = new Ajv2020({ allErrors: true, strict: false });
  const validate = ajv.compile(schema);
  return (data: unknown): void => {
    if (!validate(data)) {
      const msg = validate.errors?.map((e: any) => `${e.instancePath} ${e.message}`).join('; ') ?? 'validation failed';
      throw new ValidationError(msg, validate.errors ?? undefined);
    }
  };
}

export const validatePassport = makeValidator('agent-passport.schema.json');
export const validateReceipt = makeValidator('work-receipt.schema.json');
export const validateEnvelope = makeValidator('security-envelope.schema.json');
export const validateDNA = makeValidator('dna.schema.json');
